drivers/entropy/entropy_ll_nrfxlib.c

```diff
@@ -0,0 +1,143 @@
+/*
+ * Copyright (c) 2019 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+#include <entropy.h>
+#include <soc.h>
+#include <device.h>
+#include <irq.h>
+#include <kernel.h>
+#include <kernel_includes.h>
+#include <ble_controller.h>
+#include <ble_controller_soc.h>
+
+#include <multithreading_lock.h>
+
+#define DEV_DATA(dev) \
+	((struct rng_driver_data *)(dev)->driver_data)
+
+/* The RNG driver data. */
+struct rng_driver_data {
+	/* Used to fill the pending client buffer with new random values after
+	 * RNG IRQ is triggered.
+	 */
+	struct k_sem sem_sync;
+};
+
+static struct rng_driver_data rng_data;
+
+static int rng_driver_get_entropy(struct device *dev, u8_t *buf, u16_t len)
+{
+	__ASSERT_NO_MSG(&rng_data == DEV_DATA(dev));
+
+	u8_t *p_dst = buf;
+	u16_t bytes_left = len;
+
+	while (bytes_left > 0) {
+		s32_t bytes_read = 0;
+
+		while (bytes_read <= 0) {
+			s32_t errcode = MULTITHREADING_LOCK_ACQUIRE();
+
+			if (errcode) {
+				return errcode;
+			}
+			bytes_read = ble_controller_rand_vector_get(p_dst,
+								    bytes_left);
+			MULTITHREADING_LOCK_RELEASE();
+
+			if (!bytes_read) {
+				/* Put the thread on wait until next interrupt
+				 * to get more random values.
+				 */
+				k_sem_take(&rng_data.sem_sync, K_FOREVER);
+			}
+		}
+
+		p_dst += bytes_read;
+		bytes_left -= bytes_read;
+	}
+
+	return 0;
+}
+
+static int rng_driver_get_entropy_isr(struct device *dev, u8_t *buf, u16_t len,
+				      u32_t flags)
+{
+	__ASSERT_NO_MSG(&rng_data == DEV_DATA(dev));
+	s32_t errcode;
+
+	if (likely((flags & ENTROPY_BUSYWAIT) == 0)) {
+		errcode = MULTITHREADING_LOCK_ACQUIRE_NO_WAIT();
```

Comment:
You cannot use k_sem_take() (well hidden below MULTITHREADING_LOCK_ACQUIRE_NO_WAIT()) in ISR context.

Reply:
Why not?
From the documentation:
"A semaphore may be given by a thread or an ISR. "

Reply:
https://docs.zephyrproject.org/latest/reference/kernel/synchronization/semaphores.html#_CPPv310k_sem_takeP5k_sem5s32_t

> Can be called by ISRs, but timeout must be set to K_NO_WAIT.

The documentation for `k_mutex_lock()` does not specify anything about ISRs. Which is it?

Reply:
Ok, good. use semaphores then. It seems general documentation and header file documentation is not completely in sync.

Documentation says this about mutex:
> Note
> Mutex objects are not designed for use by ISRs.